### ISSUE TITLE
Testing

### ISSUE THREAD

**Issue by besarthoxhaj** [besarthoxhaj/verbose-memory#1]
Description
Given an ads policy that globally prohibits ads related to the sale of drugs (legal or illegal), design an ML system that prevents these ads from being shown to users

Question Statement
Note that this is a simplification of the real Meta ads policy, which has certain exemptions and permissions.

Problem exploration
Good candidates will realize that the truth signal is poorly defined and will press for clarification:
Are pharmacies allowed to advertiser their businesses (yes)
Are they allowed to mention that they are pharmacies (yes)
Are they allowed to promote the fact that they sell medicines (no)
Are they allowed to direct you to ...
... their website homepage (yes)
... a product listing for pharmaceuticals (no)
... a product listing for something else that has menu links to their pharmacy section (yes)
... a product listing for something else that has ads for pharmaceuticals on the page (yes/no depending on your mood)
The important thing to come out of this is that semantically, the truth signal depends on the ad content and the landing page, with advertiser identity signals being predictive but not causal. We need a classifier that can distinguish between indirect mentioning something and promoting it and can detect the primary content and purpose of the landing page.
The distinction between good actors and bad actors and the realization that the problem is partially adversarial normally comes later. If they bring it up this early, you can talk about bad actor behavior, like repeatedly entering ads until they find one that passes review, continually setting up new accounts as old ones get blocked, using coded language that only their customers understand.
Good candidates will realize that:
For bad actors, we need to be looking for signs of deception (e.g., significant content mismatches)
The landing page will be really important, since there are limits to what they can hide while still facilitating their activity
Advertiser identity signals will become really important
Exceptional candidates will realize that there's not actually much overlap here and consider making completely independent models.
Reality of how candidates perform
Almost no candidate ever actually does this exploration without being pushed in the right direction. If they manage anything near this clear structuring of the problem, it's a huge positive signal.
General positive signals
The candidate realizes that there is significant subtlety in the distinctions that they are trying to make and does some exploration of this subtlety (like talk about the difficulty of distinguishing herbal medicines from real pharmaceuticals)
The candidate realizes that there are good and bad actors and they need to be treated differently
Huge positive signal: the candidate independently realizes the importance of the landing page
General negative signals
The candidate wants to jump straight into discussion of features without first considering the classes they are trying to separate
This is particularly bad if the features they want to talk about are, in fact, irrelevant, like a discussion of the characteristics of the user seeing the ad
System architecture
The main choice here is about when you block the ads -- at creation time or serving time.
All candidates should be able to intuit that the latency budget at serving time is going to be constraining, while we have greatly more inference time available at creation time. This should skew them heavily towards creation time, given the difficulty of the classification. Other pertinent observations:
There is basically no additional data available at serving time that is not available at creation time (with the possible exception of a bad actor swapping out the landing page)
Most serving time inference would be redundant -- it would simply be a repeat of previous inference
The policy does not reference the user and the user identity is likely only weakly predictive of the ads being violating or not
Creation time allows you to give the advertiser feedback and the opportunity to fix their ad
All of these things skew towards creation time blocking. About the only strong justification for doing serving-time blocking would be if you wanted to protect some user segment (like children) more strongly than others. If the candidate gets to this point, it's quite a strong signal that they know what they're doing and you can let them go on with a serving time solution. Few candidates get to this level of clarity.
Training data
All candidates need to end up at the point where they have some kind of human evaluation based system. You cannot escape it in this question. If they move to user proxy-features, tell them that these are not reliable enough, because users don't know about the policy. Candidate are typically very reluctant to discuss humans in the loop (I think they're put off by the scale of Meta), so once they mention human evaluation (even as a bootstrap), I generally tell them that they have capacity to human evaluate 1% of the ads entering the system.
All candidates should talk about the unbalanced data set and strategies for balancing it, and the implications of not doing so.
Good candidates should talk about biasing the sampling towards existing model uncertainty -- models train best on samples that they are currently misclassifying.
Features
All candidates should be able to break their features down in to those describing the ad and those describing the advertiser (and for serving time blocking, those describing the user). Few candidates actually realize the importance of the landing page, so this often goes undiscussed, if a candidate talks about the landing page, this would be strongly positive.
Features is the section where the candidate has most freedom to talk about what they know best. Analysing the ad skews heavily towards understanding images and video, so candidates should be talking embeddings, CNNs and neural nets. For video, they can be talking about LSTM and its interactions with streams. Candidates can describe techniques of text analysis and language processing for text features. The problem lends itself to transfer learning and you can ask how they would use an existing neural net that is used for photo tagging in their model.
Negative signals:
The candidate doesn't really have a good idea of the distinctions they are trying to make
The candidate gives a long list of weaker features, typically relating to campaign metadata (like targeting). A good question here is, "if an advertiser changes the user targeting of an ad that has been blocked, do you actually want to change its approval status? Has it actually changed from being violating to non-violating"
The candidate can only superficially describe anything. Most candidates have some area of deep knowledge.
Candidates do not appear to know (even superficially) about things like transfer learning or embeddings
Positive signals
Deep discussion of some area of embedding feature data for the model
A discussion of causal vs non-causal features (i.e., advertiser features)
Can an ad ever be violating for one advertiser and not another (no)? Does your model have this property?
Realization that features designed for fighting bad actors can have fairness problems (like a tendency to block new accounts aggressively)
Model structure
Model structure is an area where if the candidate is an expert in DNN design, you can let them talk substantially about that. Otherwise, a good topic of conversation is the distinction between good and bad actors:
Good actors
Everything moves really slowly (the model can be updated infrequently)
Truth signals are stable and reliable
Are non adversarial - you don't need to worry about the loop that goes through advertiser behavior
Main challenge is making subtle distinctions in content classification
Bad actors
Move very quickly (the model will need to be fresh)
Truth signals change and are not historically reliable (raters could have misclassified historical samples)
Are adversarial - actors will be trying to exploit your local minima
Main challenge is in detecting deception
These substantially different requirements have implications for the model structure. Questions you can ask:
How would you cope with the very different requirements for freshness and updating?
How much benefit do you get from training these in the same model?
Possible solutions:
Independent models (trained quickly) that act as inputs to a main model
e.g., a fraud detection model that gives its output to the main classifier
Partial training of some parts of the model without training other parts
Totally separate models
All candidates should know that their model is going to terminate in some kind of classification layer (like logistic or softmax) and be able to describe the properties of it.
Evaluation
Good candidates should realize this is a precision/recall tradeoff and that we want to bias towards recall. All candidates should be able to describe a holdout strategy. All candidates should mention online/offline evaluation. All candidates should mention metrics like AuC.
Good candidates should be able to describe why online and offline metrics might differ (bad actors responding to the model). Good candidates should be able to describe the properties of their chosen metric and how it responds to things like the biased sampling in the training data generation stage.
